src/uct/tcp/tcp_ep.c

```diff
@@ -62,64 +61,203 @@ static void uct_tcp_ep_epoll_ctl(uct_tcp_ep_t *ep, int op)
 
 static inline int uct_tcp_ep_can_send(uct_tcp_ep_t *ep)
 {
-    ucs_assert(ep->offset <= ep->length);
+    ucs_assert(ep->tx.offset <= ep->tx.length);
     /* TODO optimize to allow partial sends/message coalescing */
-    return ep->length == 0;
+    return ep->tx.length == 0;
+}
+
+static inline void uct_tcp_ep_ctx_rewind(uct_tcp_ep_ctx_t *ctx)
+{
+    ctx->offset = 0;
+    ctx->length = 0;
+}
+
+static ucs_status_t uct_tcp_ep_ctx_init(uct_tcp_iface_t *iface,
+                                        uct_tcp_ep_ctx_t *ctx)
+{
+    ucs_assertv(ctx->buf == NULL, "ctx=%p", ctx);
+
+    ctx->buf = ucs_malloc(ucs_max(iface->config.buf_size,
+                                  iface->config.short_size), "tcp_buf");
+    if (ctx->buf == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    uct_tcp_ep_ctx_rewind(ctx);
+
+    return UCS_OK;
+}
+
+static void uct_tcp_ep_ctx_cleanup(uct_tcp_ep_ctx_t *ctx)
+{
+    if (ctx->buf != NULL) {
+        ucs_free(ctx->buf);
+        ctx->buf = NULL;
+    }
+}
+
+void uct_tcp_ep_ctx_migrate(uct_tcp_ep_ctx_t *to_ctx, uct_tcp_ep_ctx_t *from_ctx)
+{
+    ucs_assertv(to_ctx->buf == NULL, "to_ctx=%p", to_ctx);
+    ucs_assertv(from_ctx->buf != NULL, "from_ctx=%p", from_ctx);
+
+    *to_ctx = *from_ctx;
+
+    memset(from_ctx, 0, sizeof(*from_ctx));
+}
+
+static void uct_tcp_ep_addr_cleanup(ucs_sock_addr_t *sock_addr)
+{
+    if (sock_addr->addr != NULL) {
+        ucs_free((void*)sock_addr->addr);
+    }
+
+    memset(sock_addr, 0, sizeof(*sock_addr));
+}
+
+static ucs_status_t uct_tcp_ep_addr_init(ucs_sock_addr_t *sock_addr,
+                                         const struct sockaddr *addr)
+{
+    socklen_t addr_len = sizeof(struct sockaddr_in); /* TODO */
+    struct sockaddr *new_addr;
+
+    if (addr == NULL) {
+        memset(sock_addr, 0, sizeof(*sock_addr));
+    } else {
+        new_addr = ucs_malloc(addr_len, "sock_addr");
+        if (new_addr == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        sock_addr->addr    = new_addr;
+        sock_addr->addrlen = addr_len;
+    }
+
+    return UCS_OK;
+}
+
+static void uct_tcp_ep_close_fd(int *fd_p)
+{
+    if (*fd_p != -1) {
+        close(*fd_p);
+        *fd_p = -1;
+    }
+}
+
+static unsigned uct_tcp_ep_in_iface(uct_tcp_ep_t *ep)
+{
+    return !ucs_list_is_empty(&ep->list);
```

Comment:
i don't think it works; ep->list is not list head, it's list element. ucs_list_del() does not change the contents of &ep->list to make it look like empty list head.

Reply:
we don't need to handle this case - if we need it, we should do `ucs_list_del()` -> `ucs_list_head_init()`

we call `ucs_list_head_init()` for `&ep->list` during initialization and this handles a situation when an EP object was not inserted into the list. This is needed for further PRs to destroy an EP created during `accept()`